plan/point_select_plan.go

```diff
@@ -0,0 +1,435 @@
+		if !pm.RequestVerification(dbName, fastPlan.TblInfo.Name.L, "", checkType) {
+			return errors.New("privilege check fail")
+		}
+	}
+	return nil
+}
+
+func buildSchemaFromFields(ctx sessionctx.Context, tbl *model.TableInfo, fields []*ast.SelectField) *expression.Schema {
+	dbName := model.NewCIStr(ctx.GetSessionVars().CurrentDB)
+	columns := make([]*expression.Column, 0, len(tbl.Columns)+1)
+	if len(fields) == 1 && fields[0].WildCard != nil {
+		for _, col := range tbl.Columns {
+			columns = append(columns, colInfoToColumn(dbName, tbl.Name, col, len(columns)))
+		}
+		return expression.NewSchema(columns...)
+	}
+	if len(fields) > 0 {
+		for _, field := range fields {
+			colNameExpr, ok := field.Expr.(*ast.ColumnNameExpr)
+			if !ok {
+				return nil
+			}
+			col := findCol(tbl, colNameExpr.Name)
+			if col == nil {
+				return nil
+			}
+			columns = append(columns, colInfoToColumn(dbName, tbl.Name, col, len(columns)))
+		}
+		return expression.NewSchema(columns...)
+	}
+	// fields len is 0 for update and delete.
+	var handleCol *expression.Column
+	for _, col := range tbl.Columns {
+		column := colInfoToColumn(dbName, tbl.Name, col, len(columns))
+		if tbl.PKIsHandle && mysql.HasPriKeyFlag(col.Flag) {
+			handleCol = column
+		}
+		columns = append(columns, column)
+	}
+	if handleCol == nil {
+		handleCol = colInfoToColumn(dbName, tbl.Name, model.NewExtraHandleColInfo(), len(columns))
+		columns = append(columns, handleCol)
+	}
+	schema := expression.NewSchema(columns...)
+	schema.TblID2Handle = make(map[int64][]*expression.Column)
+	schema.TblID2Handle[tbl.ID] = []*expression.Column{handleCol}
+	return schema
+}
+
+func getSingleTableName(tableRefs *ast.TableRefsClause) *ast.TableName {
+	if tableRefs == nil || tableRefs.TableRefs == nil || tableRefs.TableRefs.Right != nil {
+		return nil
+	}
+	tblSrc, ok := tableRefs.TableRefs.Left.(*ast.TableSource)
+	if !ok {
+		return nil
+	}
+	if tblSrc.AsName.L != "" {
+		return nil
+	}
+	tblName, ok := tblSrc.Source.(*ast.TableName)
+	if !ok {
+		return nil
+	}
+	return tblName
+}
+
+func getNameValuePairs(nvPairs []nameValuePair, expr ast.ExprNode) []nameValuePair {
+	binOp, ok := expr.(*ast.BinaryOperationExpr)
+	if !ok {
+		return nil
+	}
+	if binOp.Op == opcode.LogicAnd {
+		nvPairs = getNameValuePairs(nvPairs, binOp.L)
+		if nvPairs == nil {
+			return nil
+		}
+		nvPairs = getNameValuePairs(nvPairs, binOp.R)
+		if nvPairs == nil {
+			return nil
+		}
+		return nvPairs
+	} else if binOp.Op == opcode.EQ {
+		colName, ok := binOp.L.(*ast.ColumnNameExpr)
+		if !ok {
+			return nil
+		}
+		var d types.Datum
+		switch x := binOp.R.(type) {
+		case *ast.ValueExpr:
+			d = x.Datum
+		case *ast.ParamMarkerExpr:
+			d = x.Datum
+		}
+		if d.IsNull() {
+			return nil
+		}
+		return append(nvPairs, nameValuePair{colName: colName.Name.Name.L, value: d})
+	}
+	return nil
+}
+
+func findPKHandle(tblInfo *model.TableInfo, pairs []nameValuePair) (d types.Datum) {
+	if !tblInfo.PKIsHandle {
+		return d
```

Comment:
d.SetNULL or return something like NULL constant is bettter
This code come with the assumption `types.KindInt64 != 0`

Reply:
uninitialized Datum is null is a very intuitive and reasonable assumption.